src/transformStateWithClones.js

```diff
@@ -8,6 +8,30 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  let newState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        newState = { ...newState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+        break;
+
+      case 'clear':
+        newState = {};
+        break;
+    }
+
+    stateHistory.push({ ...newState });
```

Comment:
Instead of creating a new copy of 'newState' on each iteration, you could push the state to 'stateHistory' after each action is processed. This would avoid unnecessary copying of the state object when no changes are made.

Reply:
You could simply pass newState without using the spread on it. Why? Because you have already created newState using the spread operator, there's no need to create another new object here.